### PR TITLE
SWAT-1020 -- Adding new 'ready' function/Promise interface

### PR DESCRIPTION
**Pull request by Velveeta**
[SWAT-1020](https://bits.bazaarvoice.com/jira/browse/SWAT-1020)

Per a discussion yesterday, it would be nice to be able to queue up callbacks for application classes, which wouldn't fire until the application was deemed 'loaded' and available for interacting with. The intent is that applications loaded via BV Loader should automatically kick off the callback queue once they're finished being loaded.

After some discussing about the interface, we came to an agreement that in environments where ES6 Promises were not available, `ready` will behave as a simple callback aggregator, and return undefined: e.g. `app.ready(callback1); app.ready(callbackN);`. In environments where Promises *are* available, either due to native support, or polyfilled support, it will provide the same callback aggregation, but also return a Promise that can be used to queue up more callbacks (or the Promise interface could be used exclusively), e.g. `app.ready(callback1).then(callback2); app.ready().then(callbackN);`

Callbacks will be given an instance of the loaded application, and when passed to the `ready` function itself, should follow a standard Node errback function signature (err, app), versus when passed via a Promise chain, should split those parameters for their `then` and `catch` branches, e.g. `app.ready((err, app) => {})` or `app.ready().then((app) => {}).catch((err) => {})`.

### CODE REVIEW

**Comment by Velveeta** [comment]
Well, the design *is* intentional, and the reasoning was that if a client is supporting a mix of browsers, some of which may and some of which may not have Promise support, then they should be polyfilling their own environment to account for that, if they're planning to use Promise-style syntax. If they're not polyfilling, and are trying to use the Promise-style interface, and it breaks on some legacy browser they support, that's a problem that's technically on their end, because they're the ones consuming the code in a Promise fashion without polyfilling for it. They can just as easily use the `ready(callback)` syntax, and have it work in both Promise and non-Promise environments, but if they support non-Promise-compliant browsers, and are using the Promise-style interface without polyfilling for it on their end, then that's an implementation issue on their end.

**Comment by savetheclocktower** [comment]
Sure. My only point — lodged not as a serious reservation, but just as something worth thinking about — is whether that will create support work for us down the road. The answer is probably unknowable.

**Comment by Velveeta** [comment]
It's certainly a valid concern, and definitely isn't one that I know the answer to. The times we've polyfilled Promise ourselves are only when we're using our own Promises internally, not making Promise interfaces available externally for other people to consume. That being said, we had a pretty good discussion about this, and I personally feel pretty confident in this decision, but I'm going to ping @ceejtron, @tnunamak, and @aterranova-bv since they were the major contributors to that conversation yesterday, and see if they have any additional concerns around this in the context of potentially making more support work for ourselves later down the line by making this interface environment-dependent.

**Comment by tnunamak** [comment]
I've been thinking about this and TBH, the pattern feels weird to me, but leaving polyfilling up to the user of bv-ui-core makes a lot of sense. I think we should at least make this very clear by bolding some of the documentation to call attention to it, or something.

File: lib/application/README.md [hunk 1]
```diff
@@ -49,6 +69,17 @@ APP.exampleApp.render({ productId : 1 });
 
 ## Methods to Be Called by Core Application Code
 
+- `processReady(error)`: A method that your core application code can call in
+  order to access calls to `ready` that have occurred. The only parameter it
+  accepts is an optional one, for an error to be passed to the callbacks in
+  its queue, and to reject its Promise with, if used in an environment that
+  supports ES6 Promises. It will then replace the original definition of the
+  `ready` method. Queued items are processed synchronously. If your
+  application is being loaded via the BV Loader, the Loader will handle
```
Comment (tnunamak) [inline lib/application/README.md]: Maybe a bit pedantic, but we shouldn't be creating references to BV Loader or other internal BV things in this repo, which is OSS.

File: lib/application/index.js [hunk 2]
```diff
@@ -54,13 +56,38 @@ function processQueue (queue, fn, sync) {
...
 function Application (config) {
   this.config = config;
+  this._readyQueue = [];
   this._renderQueue = [];
   this._configQueue = [];
+  if (typeof global.Promise === 'function') {
+    this._promise = new Promise(function (resolve, reject) {
+      this._resolve = resolve;
+      this._reject = reject;
+    }.bind(this));
```
Comment (tnunamak) [inline lib/application/index.js]: Is this a work-around to not having arrow functions?

Comment (Velveeta) [inline lib/application/index.js]: Yep! :D

**Comment by sirugh** [comment]
FYI the API module in this library also uses fetch and promises and expects the user of the module to polyfill as necessary.

**Comment by Velveeta** [comment]
So, if that's an assumption that the library is making, that a consumer will have polyfilled for promises in order to use the api module, do we want to scrap the *check* for Promise support in this one and also just assume that promises are available, and provide the callback and promise interface as the standard?

**Comment by sirugh** [comment]
>  do we want to scrap the check for Promise support in this one and also just assume that promises are available

Yep, and call it out in the readme for this module

https://github.com/bazaarvoice/bv-ui-core/tree/master/lib/api#requirements

**Comment by tnunamak** [comment]
I suppose worst case, if anyone is currently using this module (Firebird?, Curations?) and not already polyfilling Promise, they can upgrade to these changes and either not use .ready() or add a polyfill.

**Pull request merged**